File: jobhunt/reports.py
"""Markdown report generation, emulating the artex_job_hunt output files.

Per user, per run we write to ``data/reports/<user_id>/``:

- ``YYYY-MM-DD.md``        — daily digest: funnel, TL;DR, top matches, queries
- ``current_listings.md``  — rolling working list grouped by first-seen date,
                             newest first, with NEW markers and match bars
"""
from __future__ import annotations

import datetime as dt
from collections import Counter
from pathlib import Path
from typing import TYPE_CHECKING

from sqlalchemy import select

from . import config
from .db import Listing, User
from .matching import match_bar

if TYPE_CHECKING:
    from .pipeline import RunResult


def user_reports_dir(user_id: int) -> Path:
    path = config.REPORTS_DIR / str(user_id)
    path.mkdir(parents=True, exist_ok=True)
    return path


def render_digest(user: User, result: "RunResult", top: list[Listing]) -> str:
    lines = [
        f"# Job Hunt - {result.date}",
        "",
        "**Run mode:** deterministic pipeline (no LLM)",
        f"**Funnel:** {result.funnel}",
        f"**Dropped — out of range:** {result.fetched - result.in_range} | "
        f"**Dropped — no resume fit:** {result.in_range - result.relevant}",
        "",
        "## TL;DR",
    ]
    if top:
        best = top[0]
        lines.append(
            f"- Best match: **{best.title}** at {best.company} — "
            f"{match_bar(best.score)} ({best.location})."
        )
        lines.append(f"- {result.new_count} new listing(s) in this digest.")
        source_counts = Counter(l.source for l in top)
        lines.append(
            "- Sources represented in top matches: "
            + ", ".join(f"{s}: {n}" for s, n in source_counts.most_common())
        )
    else:
        lines.append("- No new matching listings this run.")
    lines += ["", "## Top matches"]
    for i, listing in enumerate(top, 1):
        lines += [
            f"### {i}. {listing.title} - {listing.company} ({listing.location})",
            f"- **Match:** {match_bar(listing.score)}",
            f"- **Posted:** {listing.posted or 'unknown'} | "
            f"**Source:** {listing.source}",
            f"- **Apply:** {listing.url}",
            "",
        ]
    if result.errors:
        lines += ["## Warnings"]
        lines += [f"- {e}" for e in result.errors]
        lines.append("")
    lines += ["## Queries fanned"]
    lines += [f"- {q}" for q in result.queries]
    return "\n".join(lines) + "\n"


def render_current_listings(user: User, listings: list[Listing], now: str) -> str:
    lines = [
        "# Job Hunt — Working List",
        "",
        f"_Last refreshed: {now}_",
        f"_Active listings: **{len(listings)}** "
        f"({config.RETENTION_DAYS}-day retention)_",
        "",
        "Sorted by date first seen, newest first. "
        "**NEW** = surfaced in the most recent run.",
        "",
    ]
    current_date = None
    for listing in listings:
        if listing.first_seen != current_date:
            current_date = listing.first_seen
            lines += [f"## {current_date}", ""]
        marker = "**🆕 NEW** — " if listing.is_new else ""
        lines += [
            f"### {marker}{listing.title} — {listing.company}",
            f"- **Match:** {match_bar(listing.score)}",
            f"- **Location:** {listing.location or 'unknown'} | "
            f"**Source:** {listing.source} | **Posted:** {listing.posted or 'unknown'}",
            f"- **Apply:** {listing.url}",
            "",
        ]
    return "\n".join(lines) + "\n"


def write_reports(db, user: User, result: "RunResult") -> Path:
    """Write both report files; returns the digest path."""
    top = sorted(result.new_listings, key=lambda l: l.score, reverse=True)[:10]
    digest = render_digest(user, result, top)

    listings = db.scalars(
        select(Listing)
        .where(Listing.user_id == user.id, Listing.status != "hidden")
        .order_by(Listing.first_seen.desc(), Listing.score.desc())
    ).all()
    now = dt.datetime.now().strftime("%Y-%m-%d %H:%M")
    working_list = render_current_listings(user, listings, now)

    out_dir = user_reports_dir(user.id)
    digest_path = out_dir / f"{result.date}.md"
    digest_path.write_text(digest, encoding="utf-8")
    (out_dir / "current_listings.md").write_text(working_list, encoding="utf-8")
    return digest_path

File: android/core/src/main/kotlin/com/jobhunt/core/Reports.kt
package com.jobhunt.core

/**
 * Markdown report rendering, matching the original cron's output so digests
 * stay diffable and readable outside the app (share sheet, Obsidian, Drive).
 */
object Reports {

    /** The per-run digest: funnel, TL;DR, top matches, and the queries fanned. */
    fun renderDigest(result: RunResult): String = buildString {
        appendLine("# Job Hunt - ${result.date}")
        appendLine()
        appendLine("**Run mode:** deterministic on-device pipeline (no LLM)")
        appendLine("**Funnel:** ${result.funnel}")
        appendLine(
            "**Dropped — out of range:** ${result.fetched - result.inRange} | " +
                "**Dropped — no resume fit:** ${result.inRange - result.relevant}",
        )
        appendLine()
        appendLine("## TL;DR")

        val top = result.newListings.take(10)
        val best = top.firstOrNull()
        if (best != null) {
            appendLine(
                "- Best match: **${best.title}** at ${best.company} — " +
                    "${Matching.matchBar(best.score)} (${best.location}).",
            )
            appendLine("- ${result.newCount} new listing(s) in this digest.")
            val bySource = top.groupingBy { it.source }.eachCount()
                .entries.sortedByDescending { it.value }
                .joinToString(", ") { "${it.key}: ${it.value}" }
            appendLine("- Sources represented in top matches: $bySource")
        } else {
            appendLine("- No new matching listings this run.")
        }

        appendLine()
        appendLine("## Top matches")
        top.forEachIndexed { index, listing ->
            appendLine(
                "### ${index + 1}. ${listing.title} - ${listing.company} " +
                    "(${listing.location})",
            )
            appendLine("- **Match:** ${Matching.matchBar(listing.score)}")
            appendLine(
                "- **Posted:** ${listing.posted.ifBlank { "unknown" }} | " +
                    "**Source:** ${listing.source}",
            )
            appendLine("- **Apply:** ${listing.url}")
            appendLine()
        }

        if (result.errors.isNotEmpty()) {
            appendLine("## Warnings")
            result.errors.forEach { appendLine("- $it") }
            appendLine()
        }

        appendLine("## Queries fanned")
        result.queries.forEach { appendLine("- $it") }
    }

    /** The rolling working list, grouped by first-seen date, newest first. */
    fun renderWorkingList(
        listings: List<Listing>,
        refreshedAt: String,
        retentionDays: Int = RETENTION_DAYS,
    ): String = buildString {
        appendLine("# Job Hunt — Working List")
        appendLine()
        appendLine("_Last refreshed: ${refreshedAt}_")
        appendLine("_Active listings: **${listings.size}** ($retentionDays-day retention)_")
        appendLine()
        appendLine(
            "Sorted by date first seen, newest first. " +
                "**NEW** = surfaced in the most recent run.",
        )
        appendLine()

        var currentDate: String? = null
        for (listing in listings) {
            if (listing.firstSeen != currentDate) {
                currentDate = listing.firstSeen
                appendLine("## $currentDate")
                appendLine()
            }
            val marker = if (listing.isNew) "**🆕 NEW** — " else ""
            appendLine("### $marker${listing.title} — ${listing.company}")
            appendLine("- **Match:** ${Matching.matchBar(listing.score)}")
            appendLine(
                "- **Location:** ${listing.location.ifBlank { "unknown" }} | " +
                    "**Source:** ${listing.source} | " +
                    "**Posted:** ${listing.posted.ifBlank { "unknown" }}",
            )
            appendLine("- **Apply:** ${listing.url}")
            appendLine()
        }
    }
}
